### PR TITLE
Minor refactoring/niceties for upcoming polycone work

### PR DESCRIPTION
**Pull request by sethrj**
- Add structured binding support to `celeritas::array` by implementing tuple methods
- Move `monotonic_decreasing` and VectorUtils to corecel and genericize the underlying algorithm along the lines of "any of"
- Expose geometry tolerance to surface builder so that we can use "soft equivalence" for twisted faces

### CODE REVIEW

File: src/corecel/cont/Array.hh [hunk 1]
```diff
@@ -127,3 +146,23 @@ operator!=(Array<T, N> const& lhs, Array<T, N> const& rhs)
...
+struct tuple_size<celeritas::Array<T, N>>
+{
+    static constexpr std::size_t value = N;
+};
+
+//! Support structured binding: array element type
+template<std::size_t I, class T, celeritas::size_type N>
+struct tuple_element<I, celeritas::Array<T, N>>
+{
+    using type = T;
```
Comment (esseivaju) [inline src/corecel/cont/Array.hh:164]:
```suggestion
    static_assert(I < std::tuple_size<celeritas::Array<T, N>>::value);
    using type = T;
```

File: src/corecel/grid/VectorUtils.hh [hunk 2]
```diff
@@ -24,8 +26,28 @@ std::vector<double> linspace(double start, double stop, size_type n);
...
+        return left <= right;
+    });
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * True if the grid values are monotonically increasing.
+ */
+template<class T>
+bool is_monotonic_increasing(Span<T> grid)
```
Comment (esseivaju) [inline src/corecel/grid/VectorUtils.hh:45]:
```suggestion
inline bool is_monotonic_increasing(Span<T> grid)
```

**Review by elliottbiondo — APPROVED** [review]

**Review by esseivaju — COMMENTED** [review]
I had these two minor comments that I didn't publish since the CI was failing.

**Comment by sethrj** [comment]
Ah sorry bout that! Will include in follow on.

**Pull request merged**